gokart/gcs_obj_metadata_client.py

```diff
@@ -101,23 +107,44 @@ def _get_patched_obj_metadata(
         # However, users who utilize custom_labels are no longer expected to search using the labels generated from task parameters.
         # Instead, users are expected to search using the labels they provided.
         # Therefore, in the event of a key conflict, the value registered by the user-provided labels will take precedence.
-        _merged_labels = GCSObjectMetadataClient._merge_custom_labels_and_task_params_labels(normalized_task_params_labels, normalized_custom_labels)
+        normalized_labels = [normalized_custom_labels, normalized_task_params_labels]
+        if required_task_outputs:
+            normalized_labels.append({'__required_task_outputs': json.dumps(GCSObjectMetadataClient._get_serialized_string(required_task_outputs))})
+
+        _merged_labels = GCSObjectMetadataClient._merge_custom_labels_and_task_params_labels(normalized_labels)
         return GCSObjectMetadataClient._adjust_gcs_metadata_limit_size(dict(metadata) | _merged_labels)
 
+    @staticmethod
+    def _get_serialized_string(required_task_outputs: FlattenableItems[RequiredTaskOutput]) -> FlattenableItems[str]:
+        def _iterable_flatten(nested_list: Iterable) -> list[str]:
+            flattened_list: list[str] = []
+            for item in nested_list:
+                if isinstance(item, Iterable):
+                    flattened_list.extend(_iterable_flatten(item))
+                else:
+                    flattened_list.append(item)
+            return flattened_list
```

Comment:
How about using `Iterator` like below. This can be (maybe) reduce temporal memory usage and it looks a bit elegant.
```suggestion
        def _iterable_flatten(nested_list: Iterable) -> Iterator[str]:
            for item in nested_list:
                if isinstance(item, Iterable):
                    yield from _iterable_flatten(item)
                else:
                    yield item
```

and we should change L130 as
```python
return list(_iterable_flatten([GCSObjectMetadataClient._get_serialized_string(ro) for ro in required_task_outputs]))
```

Reply:
Your suggestion is so efficient about memory usage!
I'm not familiar with `yield` and `yield from`, but this only returns generator.

In the original code (the one that returns a list), even with deep nesting or a large number of elements, it tends to be inefficient because it builds and holds the entire flattened list in memory at the end, and also creates intermediate lists during the recursive calls.

This approach is better because it's more memory-efficient, especially for large datasets or deeply nested structures, as it avoids keeping the entire flattened sequence in memory simultaneously until it's all collected.

I defenitely accept this suggest, thank you!